hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java

```diff
@@ -602,7 +563,7 @@ public void removeVolumes(
         new ArrayList<>(storageLocsToRemove);
     Map<String, List<ReplicaInfo>> blkToInvalidate = new HashMap<>();
     List<String> storageToRemove = new ArrayList<>();
-    try (AutoCloseableLock lock = datasetWriteLock.acquire()) {
+    synchronized (this) {
```

Comment:
how about making the entire method synchronized instead?

Reply:
This object lock is protect data structure related volumes like add and remove.So just replace dataset global write lock.HDFS-9445 fix problem if use synchronized to whole method.